### ISSUE TITLE
FindBin.pm fails for setuid script on solaris

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4454]
Migrated from [rt.perl.org#7758](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7758) (status was 'rejected')

Searchable as RT7758$

**Comment by p5pRT** [comment]
### From psfales@ihgp.ih.lucent.com


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
When a perl script is marked setuid on recent versions of solaris\, $0
is passed in as /dev/fd/3\.  I understand that this is a feature of the
OS\, but it breaks the FindBin\.pm module\.  Perhaps there is no way to
around this\, but perhaps FindBin\.pm could at least check for this case
and produce an error message that would help explain what is going on\.
Here's the test program&#8203;:

&nbsp;        \#\!/usr/bin/perl \-w
&nbsp;        print "\\$0 = $0\\n";

&nbsp;        use FindBin qw\($Bin\);
&nbsp;        print "$Bin\\n";


When this program is run\, we get&#8203;:
&nbsp;                                   
Cannot find current script '/dev/fd/3' at
/opt/exp/perl/lib/5\.6\.1/FindBin\.pm line 166
BEGIN failed\-\-compilation aborted at /opt/exp/perl/lib/5\.6\.1/FindBin\.pm
line 16\!
6\.                                                                     
&nbsp;       
Compilation failed in require at /dev/fd/3 line 4\.
BEGIN failed\-\-compilation aborted at /dev/fd/3 line 4\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=medium

Site configuration information for perl v5.6.1:

Configured by expmake at Thu Sep 20 22:53:15 EDT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=solaris, osvers=2.5.1, archname=sun4-solaris
    uname='sunos orion 5.5.1 generic_103640-29 sun4u sparc '
    config_args='-d'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='',
    optimize='-O',
    cppflags=''
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/opt/exp/perl/lib -L/usr/ucblib -R/usr/ucblib'
    libpth=/opt/exp/perl/lib /lib /usr/lib /usr/ccs/lib /usr/ucblib
    libs=-lsocket -lnsl -ldbm -ldl -lm -lc -lcrypt -lsec -lucb
    perllibs=-lsocket -lnsl -ldl -lm -lc -lcrypt -lsec -lucb
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags='-G -L/opt/exp/perl/lib -L/usr/ucblib -R/usr/ucblib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /opt/exp/perl/lib/5.6.1/sun4-solaris
    /opt/exp/perl/lib/5.6.1
    /opt/exp/perl/lib/site_perl/5.6.1/sun4-solaris
    /opt/exp/perl/lib/site_perl/5.6.1
    /opt/exp/perl/lib/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/psfales
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/psfales/bin:/home/psfales/bin:/opt/exptools/bin:/opt/dwb/bin:/usr/bin:/usr/sbin:/opt/bin:/opt/SUNWspro/bin:/usr/ccs/bin:/opt/ssh/bin:/opt/sudo/bin::/opt/unison/bin
    PERL_BADLANG (unset)
    SHELL=/bin/ksh93

```
</details>

**Comment by p5pRT** [comment]
### From @Hugmeir

On Thu Sep 27 08&#8203;:23&#8203;:30 2001\, psfales@&#8203;ihgp\.ih\.lucent\.com wrote&#8203;:
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> When a perl script is marked setuid on recent versions of solaris\, $0
> is passed in as /dev/fd/3\.  I understand that this is a feature of the
> OS\, but it breaks the FindBin\.pm module\.  Perhaps there is no way to
> around this\, but perhaps FindBin\.pm could at least check for this case
> and produce an error message that would help explain what is going on\.
> Here's the test program&#8203;:
> 
>         \#\!/usr/bin/perl \-w
>         print "\\$0 = $0\\n";
> 
>         use FindBin qw\($Bin\);
>         print "$Bin\\n";
> 
> 
> When this program is run\, we get&#8203;:
> 
> Cannot find current script '/dev/fd/3' at
> /opt/exp/perl/lib/5\.6\.1/FindBin\.pm line 166
> BEGIN failed\-\-compilation aborted at
> /opt/exp/perl/lib/5\.6\.1/FindBin\.pm
> line 16\!
> 6\.
> 
> Compilation failed in require at /dev/fd/3 line 4\.
> BEGIN failed\-\-compilation aborted at /dev/fd/3 line 4\.
> 
> 

Still present in 5\.16\.0\. Should this be marked as Won't Fix?

Tangentially related\, it's worth pointing out that this bug wasn't
entirely abandoned;
http&#8203;://www\.nntp\.perl\.org/group/perl\.perl5\.porters/2001/09/msg44435\.html
shows that there were at least two replies\, which I assume weren't
imported into rt\.

\-\-hugmeir

**Comment by p5pRT** [comment]
### From @nwc10

On Sat\, May 26\, 2012 at 11&#8203;:18&#8203;:01PM \-0700\, Brian Fraser via RT wrote&#8203;:
> On Thu Sep 27 08&#8203;:23&#8203;:30 2001\, psfales@&#8203;ihgp\.ih\.lucent\.com wrote&#8203;:
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > When a perl script is marked setuid on recent versions of solaris\, $0
> > is passed in as /dev/fd/3\.  I understand that this is a feature of the
> > OS\, but it breaks the FindBin\.pm module\.  Perhaps there is no way to
> > around this\, but perhaps FindBin\.pm could at least check for this case
> > and produce an error message that would help explain what is going on\.
> > Here's the test program&#8203;:
> > 
> >         \#\!/usr/bin/perl \-w
> >         print "\\$0 = $0\\n";
> > 
> >         use FindBin qw\($Bin\);
> >         print "$Bin\\n";
> > 
> > 
> > When this program is run\, we get&#8203;:
> > 
> > Cannot find current script '/dev/fd/3' at
> > /opt/exp/perl/lib/5\.6\.1/FindBin\.pm line 166
> > BEGIN failed\-\-compilation aborted at
> > /opt/exp/perl/lib/5\.6\.1/FindBin\.pm
> > line 16\!
> > 6\.
> > 
> > Compilation failed in require at /dev/fd/3 line 4\.
> > BEGIN failed\-\-compilation aborted at /dev/fd/3 line 4\.
> > 
> > 
> 
> Still present in 5\.16\.0\. Should this be marked as Won't Fix?

I don't think that it should be a "won't fix"\, at least not until we've
exhausted exploring how to make FindBin detect the case and work better\.

I think a good enough test case is something like this&#8203;:

$ echo 'print "Hello world from $FindBin&#8203;::Bin\\n"' | \./perl \-Ilib \-MFindBin /dev/fd/0
Cannot find current script '/dev/fd/0' at lib/FindBin\.pm line 166\.
BEGIN failed\-\-compilation aborted at lib/FindBin\.pm line 166\.
Compilation failed in require\.
BEGIN failed\-\-compilation aborted\.

The problem seems to be this line in FindBin&#8203;:

&nbsp;     croak\("Cannot find current script '$0'"\) unless\(\-f $script\);


I'm not sure if the \-f is reasonable\, or whether it would be better with a
test for anything that can be read like a file\.

Or whether $script =~ m\!\\A/dev/fd/\\d\+\\z\! should be special\-cased for a
different error message\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @ap

\* Nicholas Clark \<nick@&#8203;ccl4\.org> \[2012\-06\-04 16&#8203;:45\]&#8203;:
> Or whether $script =~ m\!\\A/dev/fd/\\d\+\\z\! should be special\-cased for
> a different error message\.

Are setuid scripts even regularly supported any more?

**Comment by p5pRT** [comment]
### From @ikegami

On Sun\, May 27\, 2012 at 2&#8203;:18 AM\, Brian Fraser via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Thu Sep 27 08&#8203;:23&#8203;:30 2001\, psfales@&#8203;ihgp\.ih\.lucent\.com wrote&#8203;:
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > When a perl script is marked setuid on recent versions of solaris\, $0
> > is passed in as /dev/fd/3\.  I understand that this is a feature of the
> > OS\, but it breaks the FindBin\.pm module\.  Perhaps there is no way to
> > around this\, but perhaps FindBin\.pm could at least check for this case
> > and produce an error message that would help explain what is going on\.
> > Here's the test program&#8203;:
> >
> >         \#\!/usr/bin/perl \-w
> >         print "\\$0 = $0\\n";
> >
> >         use FindBin qw\($Bin\);
> >         print "$Bin\\n";
> >
> >
> > When this program is run\, we get&#8203;:
> >
> > Cannot find current script '/dev/fd/3' at
> > /opt/exp/perl/lib/5\.6\.1/FindBin\.pm line 166
> > BEGIN failed\-\-compilation aborted at
> > /opt/exp/perl/lib/5\.6\.1/FindBin\.pm
> > line 16\!
> > 6\.
> >
> > Compilation failed in require at /dev/fd/3 line 4\.
> > BEGIN failed\-\-compilation aborted at /dev/fd/3 line 4\.
> >
> >
>
> Still present in 5\.16\.0\. Should this be marked as Won't Fix?
>

Actually\, I'd reject for not being a bug\.

First\, returning an error is not necessarily a bug\.

Second\, the error is not the result of the bug\. Find&#8203;::Bin has no way of
knowing where the script is located based on the info the system provided
\($0="/dev/fd/3"\)\.

Third\, even if Find&#8203;::Bin did have some way of finding out the original path
to the file\, it shouldn't use it because it would defy the system's
security measure of passing /dev/fd/3 to perl\.

\- Eric

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Jun 06 15&#8203;:04&#8203;:01 2012\, ikegami@&#8203;adaelis\.com wrote&#8203;:
> On Sun\, May 27\, 2012 at 2&#8203;:18 AM\, Brian Fraser via RT \<
> perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> 
> > On Thu Sep 27 08&#8203;:23&#8203;:30 2001\, psfales@&#8203;ihgp\.ih\.lucent\.com wrote&#8203;:
> > >
> > > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > When a perl script is marked setuid on recent versions of solaris\, $0
> > > is passed in as /dev/fd/3\.  I understand that this is a feature of the
> > > OS\, but it breaks the FindBin\.pm module\.  Perhaps there is no way to
> > > around this\, but perhaps FindBin\.pm could at least check for this case
> > > and produce an error message that would help explain what is going on\.
> > > Here's the test program&#8203;:
> > >
> > >         \#\!/usr/bin/perl \-w
> > >         print "\\$0 = $0\\n";
> > >
> > >         use FindBin qw\($Bin\);
> > >         print "$Bin\\n";
> > >
> > >
> > > When this program is run\, we get&#8203;:
> > >
> > > Cannot find current script '/dev/fd/3' at
> > > /opt/exp/perl/lib/5\.6\.1/FindBin\.pm line 166
> > > BEGIN failed\-\-compilation aborted at
> > > /opt/exp/perl/lib/5\.6\.1/FindBin\.pm
> > > line 16\!
> > > 6\.
> > >
> > > Compilation failed in require at /dev/fd/3 line 4\.
> > > BEGIN failed\-\-compilation aborted at /dev/fd/3 line 4\.
> > >
> > >
> >
> > Still present in 5\.16\.0\. Should this be marked as Won't Fix?
> >
> 
> Actually\, I'd reject for not being a bug\.
> 
> First\, returning an error is not necessarily a bug\.
> 
> Second\, the error is not the result of the bug\. Find&#8203;::Bin has no way of
> knowing where the script is located based on the info the system provided
> \($0="/dev/fd/3"\)\.
> 
> Third\, even if Find&#8203;::Bin did have some way of finding out the original
path
> to the file\, it shouldn't use it because it would defy the system's
> security measure of passing /dev/fd/3 to perl\.
> 
> \- Eric


More than a year ago\, ikegami recommended rejecting this ticket as not a
bug\.  No one has disputed that recommendation since\.

I am taking this ticket for the purpose of closing it within 7 days
unless someone wants to take the issue over\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

On Wed Jun 19 16&#8203;:09&#8203;:13 2013\, jkeenan wrote&#8203;:
> On Wed Jun 06 15&#8203;:04&#8203;:01 2012\, ikegami@&#8203;adaelis\.com wrote&#8203;:
> 
> 
> More than a year ago\, ikegami recommended rejecting this ticket as not a
> bug\.  No one has disputed that recommendation since\.
> 
> I am taking this ticket for the purpose of closing it within 7 days
> unless someone wants to take the issue over\.
> 

Closing as per schedule\.

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'rejected'
